### PR TITLE
Add support for Aqara P1 motion sensor

### PR DESCRIPTION
**Pull request by dmulcahey**
**TODO:**

- [x] motion sensitivity
- [x] detection interval
- [ ] illuminance update w/o motion

Fixes: #1498

### CODE REVIEW

**Comment by MrRagga-** [comment]
Hi, I just bought the sensor. If you could tell me how to test it, I am eager to do it.
I am also familar with python development and could help a bit, too. I've just never done something for Hass/ZHA, so I would need some advise what the fastest way would be to test it.

**Comment by MrRagga-** [comment]
I just installed the quirk and I get: 
image

**Comment by MrRagga-** [comment]
Do you own a device yourself or is help appreciated? If you point me a bit to the missing parts I could support you?

**Comment by dmulcahey** [comment]
I own 2. I’m busy with the current HA beta / release. I’ll finish this up in a week or so. What’s here now works. I just need to add stuff for the configuration

**Comment by MrRagga-** [comment]
Okay. Understood. Then it does not make to help you, I guess. Thanks a lot for your work.

**Comment by karthik-krishnan** [comment]
Is it possible to change the timeout interval for detection with the quirk?

**Comment by dmulcahey** [comment]
> Is it possible to change the timeout interval for detection with the quirk?

When I’m done it will be

File: zhaquirks/xiaomi/aqara/motion_ac02.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+    def _update_attribute(self, attrid: int, value: Any) -> None:
+        super()._update_attribute(attrid, value)
+        if attrid == 274:
+            value = value - 65536
+            self.endpoint.illuminance.illuminance_reported(value)
+            self.endpoint.occupancy.update_attribute(0, 1)
+
+    async def write_attributes(
+        self, attributes: dict[str | int, Any], manufacturer: int | None = None
```
Comment (TheJulianJES) [inline zhaquirks/xiaomi/aqara/motion_ac02.py:47]: Home Assistant Core 2022.5.4/Home Assistant OS 7.6 ship with Python 3.9 but https://peps.python.org/pep-0604/ is only supported with Python 3.10.
For anyone wanting to try this out, change this line to:
```python
self, attributes: dict[Union[str, int], Any], manufacturer: Union[int, None] = None
```

Comment (TheJulianJES) [inline zhaquirks/xiaomi/aqara/motion_ac02.py:47]: Home Assistant Core 2022.5.4/Home Assistant OS 7.6 ship with Python 3.9 but https://peps.python.org/pep-0604/ is only supported with Python 3.10.
For anyone wanting to try this out, change this line to:
```python
self, attributes: dict[Union[str, int], Any], manufacturer: Union[int, None] = None
```

**Comment by TheJulianJES** [comment]
After doing a long trial and error session, attribute 338/0x0152 seems to be the one responsible for the "trigger detection LED" (flash on motion):

0: trigger indicator LED disabled
1: trigger indicator LED enabled

(For reference, see: https://github.com/Koenkk/zigbee2mqtt/issues/11990#issuecomment-1126075266)

It would be nice if that's also added

**Comment by TheJulianJES** [comment]
I know this quirk is still WIP, but not sure if this issue is known/expected:

For some reason, the light level reported at one point was very high:
Bildschirmfoto 2022-05-14 um 15 27 52

File: zhaquirks/xiaomi/aqara/motion_ac02.py (hunk 1 above)
Comment (puddly) [reply zhaquirks/xiaomi/aqara/motion_ac02.py:47]: Adding `from __future__ import annotations` as the very first import in the module will make the type annotations compatible with 3.9.

**Comment by joe81tx** [comment]
I just downloaded custom quirk for the first time, version 7c9944c. I don't see an option to modify the detection interval. If this has not yet been built into the UI is there a way to manually update the interval?

**Comment by MollyrH** [comment]
I apologize upfront for this newbie question but I have spent hours trying to research on my own without finding the answer.

I have created the custom_zha_quirks directory and added the reference in configuration.yaml.  This enabled the power/motion/presence activities.  All good.

My question is what to do with the modified  _init_.py file.  As you can see below, I added a folder under custom_components ("xiaomi" but this did not seem to do anything nor did it add the detection interval identity.

Aqara P1

In general, the Home Assistant forum is not very inviting for asking newbie questions.  I'm happy to learn if someone prefers to point me to a good source instead of just answering my question directly.

**Comment by mzdial** [comment]
I just installed my first quirk today, getting the Aqara P1 to work, thanks to @dmulcahey!

One question... Is the sensitivity something that will be adjustable in the future? (I wasn't sure if it was exposed and you could integrate.) I have this sensor on a landing in my stairwell, attempting to trigger motion either going down or up from the lower level of our home. I have to get between 1-2m to trigger it.

Thanks!

**Comment by dmulcahey** [comment]
> After doing a long trial and error session, attribute 338/0x0152 seems to be the one responsible for the "trigger detection LED" (flash on motion):
> 
> 0: trigger indicator LED disabled 1: trigger indicator LED enabled
> 
> (For reference, see: [Koenkk/zigbee2mqtt#11990 (comment)](https://github.com/Koenkk/zigbee2mqtt/issues/11990#issuecomment-1126075266))
> 
> It would be nice if that's also added

This is included

**Comment by MollyrH** [comment]
> @mzdial could you please share how you installed this quirk? At the moment, I've done this:
> 
> 1. Create a custom quirk dir in HA, e.g., /config/custom_zha_quirks
> 2. In configuration.yaml, point to this directory:
> 
> ```
> zha:
>   custom_quirks_path: /config/custom_zha_quirks/
> ```
> 
> 3. in this directory, create the python file with content of the quirk
> 
> Where do I get the files and which directory do I put them in? thanks
> 
> > I just installed my first quirk today, getting the Aqara P1 to work, thanks to @dmulcahey!
> > One question... Is the sensitivity something that will be adjustable in the future? (I wasn't sure if it was exposed and you could integrate.) I have this sensor on a landing in my stairwell, attempting to trigger motion either going down or up from the lower level of our home. I have to get between 1-2m to trigger it.
> > Thanks!

Are you asking where the quirk file is?
https://github.com/zigpy/zha-device-handlers/tree/dev/zhaquirks/xiaomi/aqara/

**Comment by mzdial** [comment]
> 

@Anto79-ops, I'm running on Docker. In my Home Assistant configuration, I'm using:

zha:
    enable_quirks: true
    custom_quirks_path: /config/custom_zha_quirks/

My HA /config is mapped to /docker/homeassistant/

Then, I downloaded the full 0.0.73 release:

[https://github.com/zigpy/zha-device-handlers/releases/tag/0.0.73](url)

I then updated the quirk file from the latest version published by @dmulcahey:

[https://github.com/zigpy/zha-device-handlers/blob/dev/zhaquirks/xiaomi/aqara/motion_ac02.py](url)

I'm pretty new to this and still learning... But this worked for me!

**Comment by MollyrH** [comment]
> 

Thanks for the input.  Its challenging to get anyone to answer newbie questions!

When you indicate you "downloaded the full .073 release", what do you mean?  I've incorporates the motion_ac02.py quirk file and can see the motion/power/presence values but I still do not see an entry to change the detection interval.

**Comment by Anto79-ops** [comment]
@MollyrH 

the link for the 0.073 release does not work for me so I just downloaded this version (updated hours ago)

https://github.com/zigpy/zha-device-handlers/tree/dev/zhaquirks/xiaomi/aqara

and can confirm I don't have the ability to change detection interval. may be a full release thins?

**Comment by MollyrH** [comment]
mzdial/Anto79

Just to confirm, all you are using/installing is the ac02.py file into the custom_zha_quirks directory and then referencing in your configuration.yaml file?  I'm asking to confirm and to also assure I know how to enable a custom quirk...

**Comment by MollyrH** [comment]
mzdial/Anto79

Did you remove and then reinstall the P1 after adding the new quirk?

**Comment by Anto79-ops** [comment]
yes. I removed the device first, delete both files __init, and ac02.py in their respective folder (custom integrations and custom zha quirks), downloaded both files from the link I provded above and put them back in the correct folders. Sucessfully added the P1 and it works well, except no interval selections.

**Comment by MollyrH** [comment]
Interesting.  I do not have a custom integrations directory nor did I install the init file.  I did not delete/re-add the P1. I  installed only the ac02.py file and restarted.  Seems to work with the exception of interval selection.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by karthik-krishnan** [comment]
Thanks for the quirk.  I have it and the fundamental motion sensing is working.  I understand the detection interval config is coming soon.  Will the quirk approach continue or is it temporary until the Core is updated with this change?

Best regards,
Karthik Krishnan
https://karthiks.in

**Comment by dmulcahey** [comment]
> @karthik-krishnan There should be a Home Assistant beta version later today (2022.6.0 beta). It should include both the quirk and configuration directly through the GUI if I'm correct.
> 
> (So the custom quirk can be removed with Home Assistant 2022.6.0 or later)

This is true

**Comment by Anto79-ops** [comment]
Hey all thanks for your time and efforts here!

I have the quirk installed now, and works great. When it gets officially released, what is good practice to swtich over to the offical release?

Do I:

1. Delete/remove the device
2. Delete the zha_custom_quirk *.py file
3. Delete the custom component  *.py file
4. Restart HA
5. Update to 2022.6.0
6. Readd device

?

**Comment by MollyrH** [comment]
updated to 2022.6.0.

Needed to delete device and re-add.

- Detection Interval setting present with default of 30 seconds (which it actually is).  Can change but will not stay at new setting,  reverts back to 30.
- Same with Sensitivity.  Defaults to Medium.  Can select Low or high but will not stay, reverts back to medium once you leave page.
- Same with cluster trigger too - reverts to off.
- Power is now unknown.

**Comment by dmulcahey** [comment]
You need to press the button on the device once immediately after changing the setting to wake the device up so it receives the command

**Comment by MollyrH** [comment]
dmulcahey,

Got it!  Works perfectly!  Thank you for all your efforts - this is awesome.

**Comment by Anto79-ops** [comment]
hey thanks again all! Interesting (but good) thing happened, after reading @MollyrH message that he had to re add the device, i just took the brute force windows approach and did this to switch out of quirk to beta release

Delete/remove the device
Delete the zha_custom_quirk *.py file
Delete the custom component *.py file
Restart HA
Update to 2022.6.0beta

At this point, I was about to add the device but it just appeared on its own...and with all new features (i.e. interval, sensitivity).

So I guess  the device goes into learning mode on its own when its removed?

**Comment by Anto79-ops** [comment]
@dmulcahey 

just wanted to bring something to you attention with regards to the illuminance (lx) number reporting in 2022.6.0beta/ZHA. I have 3 of these, two are in well lit rooms (even at night time) one is in the basement, and pretty dark most of the time.

I noticed the one in the basement, I get these spikes of 65533, 65534 lux, which corresponds to something similar in the code here:

```
 def _update_attribute(self, attrid: int, value: Any) -> None:
        super()._update_attribute(attrid, value)
        if attrid == MOTION_ATTRIBUTE:
            value = value - 65536
            self.endpoint.illuminance.illuminance_reported(value)
            self.endpoint.occupancy.update_attribute(OCCUPANCY, ON)
```

when i walk into the dark room and open a window, it detects me and updates the light from 65534 lux, to a more releaistic number to say 1.0 or 2, or 3, and I turn on a light and goes up to 27 for example, and then when I leave the room with lights back OFF, it goes back to 6553X. Now, if I leave the room with the lights ON, it behaves perfectly normal with no spikes in lx after timeout.

so it seems if the devices times out during low light detection like 1.0 lx, it results in a spike,  but if it times out on some measureable value of lx like 20 lx, it does not result in a spike after timeout

If you have one, try putting it in a dark drawer soon after it has detected something and wait for it time out.

I'm wondering if perhaps the code does not handle 0 or 1 lx well? or perhaps their is some negative numbers? and 65536 needs a bit of a tweak but should be 65540....I have no idea.

or I have a defective sensor is another option.

any thoughts?

**Comment by TheJulianJES** [comment]
@Anto79-ops The issue is already being worked on (https://github.com/zigpy/zha-device-handlers/pull/1587). That PR doesn't seem to stop the issue just yet though. Tonight I logged two invalid values: 65531 and 65535.

I'll place my P1 sensor in a dark spot and see what's the lowest invalid value.

**Comment by Anto79-ops** [comment]
Thank you @TheJulianJES for bringing this to my attention.

Yesterday the one that I thought was in the well-lit room also gave a high number when it was dark, so my theory is out the window  (pun intended).

It's almost as if the light value updates after detection has been timed out so that's kind of strange.

Hopefully it's just a software fix.

**Comment by TheJulianJES** [comment]
@Anto79-ops Yeah, the issue can probably be worked around. At the moment, there's some discussion in the Z2M repo here: https://github.com/Koenkk/zigbee2mqtt/issues/12596

I'm currently testing the sensor in various scenarios with additional logging to narrow down when the high values are actually sent.

**Comment by Anto79-ops** [comment]
Hey just to confirm that this lux issue will be fixed in the official release on June 1st?  Seems he was fixed over at the Z2M side... The dark side :)

Thanks again all for your efforts

**Comment by LCMilstein** [comment]
> hey thanks again all! Interesting (but good) thing happened, after reading @MollyrH message that he had to re add the device, i just took the brute force windows approach and did this to switch out of quirk to beta release
> 
> Delete/remove the device Delete the zha_custom_quirk *.py file Delete the custom component *.py file Restart HA Update to 2022.6.0beta
> 
> At this point, I was about to add the device but it just appeared on its own...and with all new features (i.e. interval, sensitivity).
> 
> So I guess the device goes into learning mode on its own when its removed?

Can you share a screenshot of what you're seeing? I just (a) deleted the custom quirk and everything in the custom quirk folder, updated to 2022.6.0, and then re-set-up one of these and it joined no problem, but there's no interval or sensitivity that I see. There are 2 disabled entities (basic_lqi and basic_rssi); are these related to that?

**Comment by LCMilstein** [comment]
Works!!
For anyone else trying, you actually have to remove the device and set it up like new for the new entities to show up (even if you delete the custom quirks and install a new HA version... you still have to delete the device).

Thank you for making this work!

**Comment by philibu** [comment]
Hi there - I updated with Home Assistant 2022.6.0 and see now the detection remains as "Detected" even if there is no activity in the room for quite some time. Also, I do not see how to configure the detection interval ?
Thanks

**Comment by Anto79-ops** [comment]
I think you to press the button once on the P1 (to wake it up) immediatly after changing the value in HA. Have you tried that?

> Trigger indicator, detection interval and motion sensitivity give me errors when i try to change them. I use conbee II and latest HA OS.
> 
> `Error executing service: <ServiceCall switch.turn_on (c:............): entity_id=['switch.lumi_lumi_motion_ac02_........_opple_cluster_trigger_indicator']>` `Error executing service: <ServiceCall number.set_value (c:..........): value=31.0, entity_id=['number.lumi_lumi_motion_ac02_........_opple_cluster_detection_interval']>` `Error executing service: <ServiceCall select.select_option (c:..........): option=Low, entity_id=['select.lumi_lumi_motion_ac02_........_opple_cluster_motion_sensitivity']>`
> 
> This error is from motion sensitivity:
> 
> ```
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.9/asyncio/tasks.py", line 490, in wait_for
>     return fut.result()
> asyncio.exceptions.CancelledError
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/usr/src/homeassistant/homeassistant/core.py", line 1722, in catch_exceptions
>     await coro_or_task
>   File "/usr/src/homeassistant/homeassistant/core.py", line 1741, in _execute_service
>     await cast(Callable[[ServiceCall], Awaitable[None]], handler.job.target)(
>   File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 204, in handle_service
>     await service.entity_service_call(
>   File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 680, in entity_service_call
>     future.result()  # pop exception if have
>   File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 964, in async_request_call
>     await coro
>   File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 717, in _handle_entity_call
>     await result
>   File "/usr/src/homeassistant/homeassistant/components/select/__init__.py", line 54, in async_select_option
>     await entity.async_select_option(option)
>   File "/usr/src/homeassistant/homeassistant/components/zha/select.py", line 198, in async_select_option
>     await self._channel.cluster.write_attributes(
>   File "/usr/local/lib/python3.9/site-packages/zhaquirks/xiaomi/aqara/motion_ac02.py", line 59, in write_attributes
>     result = await super().write_attributes(attributes, manufacturer)
>   File "/usr/local/lib/python3.9/site-packages/zigpy/zcl/__init__.py", line 536, in write_attributes
>     return await self.write_attributes_raw(attrs, manufacturer)
>   File "/usr/local/lib/python3.9/site-packages/zigpy/zcl/__init__.py", line 542, in write_attributes_raw
>     result = await self._write_attributes(attrs, manufacturer=manufacturer)
>   File "/usr/local/lib/python3.9/site-packages/zigpy/device.py", line 327, in request
>     return await asyncio.wait_for(req.result, timeout)
>   File "/usr/local/lib/python3.9/asyncio/tasks.py", line 492, in wait_for
>     raise exceptions.TimeoutError() from exc
> asyncio.exceptions.TimeoutError
> ```

**Comment by philibu** [comment]
Thanks. I paired it again, and this now works.
Cheers
Philippe

**Comment by jonnylangefeld** [comment]
The older, non P1 version has temperature. do you think we can add this to the P1 too?
https://github.com/zigpy/zha-device-handlers/blob/f477894d18df0b8e6c5d77e54aed8c1ae30aa956/zhaquirks/xiaomi/aqara/motion_aq2.py#L77

**Pull request merged**